src/app/util/ember-compatibility-functions.cpp

```diff
@@ -187,189 +187,215 @@ bool ServerClusterCommandExists(const ConcreteCommandPath & aCommandPath)
     return emberAfContainsServer(aCommandPath.mEndpointId, aCommandPath.mClusterId);
 }
 
-CHIP_ERROR ReadSingleClusterData(FabricIndex aAccessingFabricIndex, const ConcreteAttributePath & aPath, TLV::TLVWriter * apWriter,
-                                 bool * apDataExists)
+CHIP_ERROR ReadSingleClusterData(FabricIndex aAccessingFabricIndex, const ConcreteAttributePath & aPath,
+                                 AttributeReportIB::Builder & aAttributeReport)
 {
     ChipLogDetail(DataManagement,
                   "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%" PRIx16 " AttributeId=" ChipLogFormatMEI,
                   ChipLogValueMEI(aPath.mClusterId), aPath.mEndpointId, ChipLogValueMEI(aPath.mAttributeId));
+    AttributeDataIB::Builder attributeDataIBBuilder;
+    AttributePathIB::Builder attributePathIBBuilder;
+    AttributeStatusIB::Builder attributeStatusIBBuilder;
+    TLV::TLVWriter * writer = nullptr;
+    TLV::TLVWriter backup;
+    aAttributeReport.Checkpoint(backup);
+
+    attributeDataIBBuilder = aAttributeReport.CreateAttributeData();
+    attributePathIBBuilder = attributeDataIBBuilder.CreatePath();
+    attributePathIBBuilder.Endpoint(aPath.mEndpointId)
+        .Cluster(aPath.mClusterId)
+        .Attribute(aPath.mAttributeId)
+        .EndOfAttributePathIB();
+    ReturnErrorOnFailure(attributePathIBBuilder.GetError());
 
     AttributeAccessInterface * attrOverride = findAttributeAccessOverride(aPath.mEndpointId, aPath.mClusterId);
     if (attrOverride != nullptr)
     {
         // TODO: We should probably clone the writer and convert failures here
         // into status responses, unless our caller already does that.
-        AttributeValueEncoder valueEncoder(apWriter, aAccessingFabricIndex);
+        writer = attributeDataIBBuilder.GetWriter();
+        VerifyOrReturnError(writer != nullptr, CHIP_NO_ERROR);
+        AttributeValueEncoder valueEncoder(writer, aAccessingFabricIndex);
```

Comment:
If we now plan to guarantee that `AttributeValueEncoder` has a writer, we can simplify its code to get rid of null-checks...  Do we plan to guarantee that?